core/src/main/scala/org/bykn/bosatsu/LetFree.scala

```diff
@@ -173,89 +291,123 @@ object LetFreeConversion {
             case notMatch => notMatch
           }
         }
-      case LetFreePattern.ListPat(items) =>
-        items match {
-          case Nil =>
-            { (arg, acc) =>
-              arg match {
-                case LetFreeExpression.Struct(0, List()) => Matches(acc)
-                case LetFreeExpression.Struct(_, _) => NoMatch
-                case _ => NotProvable
-              }
+      case lp@LetFreePattern.ListPat(_) => {
+        sealed abstract class StructListResult
+        case object Empty extends StructListResult
+        case class Cons(h: T, t: StructList) extends StructListResult
```

Comment:
It's not quite a `List[T]`. `T` is something that can possibly be converted to a head and a tail; eg a `Value` or a `LetFreeExpression`. But in the case of a `LetFreeExpression`, it might not need runtime information to do the conversion and this might be running at compile time. So this structure lets me check if a pattern can be ruled in or out, even if we can't evaluate the tail. There's probably an existing way to do this in cats that I don't know about.

does that make sense?